### ISSUE TITLE
Using stable and unstable channels with flakes

### ISSUE THREAD

**Issue by timsofteng** [nix-community/home-manager#1538]
Hello. Can you help me please with multi channels issue?
I want to use packages from stable and unstable bracnches in home manager.
I'm using flakes.
How can i do it? What i need to add to my flake.nix and how I can use it inside home-manager?

Here is my flake.nix config:

`{
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-20.09";
  inputs.nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";

  inputs.home-manager.url = "github:nix-community/home-manager/master";
  inputs.home-manager.inputs.nixpkgs.follows = "/nixpkgs";

  outputs = inputs: {

    nixosConfigurations.nixos = inputs.nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      # Things in this set are passed to modules and accessible
      # in the top-level arguments (e.g. `{ pkgs, lib, inputs, ... }:`).
      specialArgs = {
        inherit inputs;
      };
      modules = [
        inputs.home-manager.nixosModules.home-manager

        ({ pkgs, ... }: {
          nix.extraOptions = "experimental-features = nix-command flakes";
          nix.package = pkgs.nixFlakes;
          nix.registry.nixpkgs.flake = inputs.nixpkgs;
          
          home-manager.useGlobalPkgs = true;

        })

        ./configuration.nix
      ];
    };

  };
}`

**Comment by jkachmar** [comment]
I don't necessarily know that this is something `home-manager` specifically concerns itself with, but I'm currently using `overlays` as a way to "inject" alternative package sets.

e.g. The following is an excerpt from my `inputs` attrset:
```nix
{
  inputs = {
    # Stable NixOS nixpkgs package set; pinned to the 20.03 release.
    nixosPkgs.url = "github:nixos/nixpkgs-channels/nixos-20.03-small";
    # Stable Darwin nixpkgs package set; pinned to the 20.03 release.
    darwinPkgs.url = "github:nixos/nixpkgs-channels/nixpkgs-20.03-darwin";
    # Tracks nixos/nixpkgs-channels unstable branch.
    #
    # Try to pull new/updated packages from 'unstable' whenever possible, as
    # these will likely have cached results from the last successful Hydra
    # jobset.
    unstable.url = "github:nixos/nixpkgs-channels/nixos-unstable";
    # Tracks nixos/nixpkgs main branch.
    #
    # Only pull from 'trunk' when channels are blocked by a Hydra jobset
    # failure or the 'unstable' channel has not otherwise updated recently for
    # some other reason.
    trunk.url = "github:nixos/nixpkgs";
  };
  # Remaining attrs elided.
}
```
...and the following is an excerpt from my `outputs` attrset:
```nix
{
  # ...
  outputs = inputs@{ self, darwinPkgs, nixosPkgs, ... }: {
    overlays = {
      # Inject 'unstable' and 'trunk' into the overridden package set, so that
      # the following overlays may access them (along with any system configs
      # that wish to do so).
      pkg-sets = (
        final: prev: {
          unstable = import inputs.unstable { system = final.system; };
          trunk = import inputs.trunk { system = final.system; };
        }
      );
    # Remaining attributes elided.
    };
}
```

This lets me reference `unstable` and `trunk` in overlays like so:
```nix
final: prev: {
  emacsMacport = final.unstable.emacsMacport;
}
```
...and in the main package set as follows:
```nix
{ pkgs, ... }:
{
  environment.systemPackages = with pkgs; [ unstable.git ];
}
```

---

What I'm doing is pulling the `system` value out of the final configuration (which avoids using something like `builtins.currentSystem`, keeping everything pure), and then using that as the bare-minimum of what's required to instantiate a `nixpkgs` package set (other config stuff like `allowUnfree` could also be applied here).

This means I can easily reference stuff from any number of additional `nixpkgs` package sets anywhere in my configuration.

The somewhat-less-than-obvious drawback is that it _seems_ to slow down evaluation (presumably this is some fixpoint nonsense) to something like 38s for my configuration, so take care about that.

**Comment by jkachmar** [comment]
I just noticed `specialArgs` in your excerpt; I don't know too much about `specialArgs` in particular, but if you don't need access to these package sets in your overlays (or if you're not using overlays at all), it might be more direct to do something like:
```nix
{
  nixosConfigurations.nixos = inputs.nixpkgs.lib.nixosSystem {
    system = "x86_64-linux";
    modules =
      let
        defaults = { pkgs, ... }: {
          _module.args.unstable = import inputs.unstable { inherit (pkgs.stdenv.targetPlatform) system; };
          _module.args.trunk = import inputs.trunk { inherit (pkgs.stdenv.targetPlatform) system; };
        };
      in [
        defaults
        inputs.home-manager.nixosModules.home-manager
        # Remaining modules elided.
      ];
};
```
...which should allow you to access `inputs` as follows from any modules imported _after_ the `_module.args.inputs` line like this:
```nix
{ pkgs, unstable ... }:
{
  environment.systemPackages = [ unstable.git ];
}
```

**Comment by berbiche** [comment]
To complement what @jkachmar said, you would need to inject your inputs within `home-manager` otherwise you would need to use an overlay.

``` nix
{
  nixosConfigurations.nixos = inputs.nixpkgs.lib.nixosSystem {
    system = "x86_64-linux";
    modules = [
      inputs.home-manager.nixosModules.home-manager
      ({ ... }: {
        home-manager.users.your-user = { ... }: {
          _module.args.inputs = inputs;
        };
      })
    ];
  };
}
```

Your inputs would the be available as a module's arguments:

``` nix
{ pkgs, inputs, ... }:

{
  # Note, `${pkgs.system}` is the "architecture" of the machine evaluating and building
  environment.systemPackages = with inputs.nixpkgs-unstable.legacyPackages.${pkgs.system}; [  ];
}
```

The logic remains valid if you are using `home-manager` outside of NixOS. You would need to inject the inputs through `_module.args`.

**Comment by teto** [comment]
Question looks answered, please reopen if not.

**Comment by badly-drawn-wizards** [comment]
Hi, I'm trying to use `_module.args`, but failing miserably.

Here is the NixOS module with `inputs` provided through specialArgs
```
{ inputs, ... }:
{
  imports = [
    inputs.home-manager.nixosModules.home-manager
  ];

  home-manager.useGlobalPkgs = true;
  home-manager.useUserPackages = true;
  home-manager.users.reuben = {
    imports = [
      {
        _module.args.inputs = inputs;
      }
      ./home-manager.nix
    ];
  };
}
```

I'm getting infinite recursion whenever I force `inputs` as an argument in a home-manager module

**Comment by berbiche** [comment]
AFAIK, an import must either be a function or a path. I think Nix is recursing into the attribute set and tries to import all inputs which causes an infinite recursion error.

IIRC this should work:

```nix
{
  home-manager.users.your-user = {
    imports = [ ./home.nix ];
    _module.args.inputs = inputs;
  };
}
```
or

```nix
{
  home-manager.users.your-user = {
    imports = [
      ({ ... }: { _module.args.inputs = inputs; })
      ./home.nix
    ];
  };
}
```

**Comment by badly-drawn-wizards** [comment]
It's seems either one gives the same result unfortunately. May be a different cause though.

**Comment by berbiche** [comment]
@badly-drawn-wizards can you open a new issue with the configuration you are making it work with?

**Comment by badly-drawn-wizards** [comment]
@berbiche Will do

**Comment by stites** [comment]
I am trying to replicate this original question but am running into some problems. Notably, in the original question I would like to have the same inputs:

```
inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-20.09";
inputs.nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";

inputs.home-manager.url = "github:nix-community/home-manager/master";
inputs.home-manager.inputs.nixpkgs.follows = "/nixpkgs-unstable";
```

(note that hm-master follows nixpkgs-unstable here)

I am able to use `_module.args` and `specialArgs` as suggested, but only when I have home-manager pinned to 20.09. Whenever i follow master, compiling flakes fails (specifically because of a `types.anything` in matplotlib).

I've tried to "swap out" with `_module.args.pkgs = unstablePkgs;` before loading the hm-module, but this doesn't seem to work. Would anyone have suggestions for this?

**Comment by berbiche** [comment]
@stites do you have `home.stateVersion = "20.09"`? Important changes were done in 20.09 regarding loading nixpkgs

**Comment by stites** [comment]
yes, I do have `home.stateVersion = "20.09"`. Another thing I just tried was to set:
```
  home-manager.useGlobalPkgs = false;
  home-manager.useUserPackages = true;
```
with the above flake inputs. this also produces an error with the missing `types.anything` -- it's not clear to me where the input is going since there are no listed inputs for nixpkgs in the flake (sorry, also a flake noob here, so maybe nixpkgs is always implied?).

**Comment by berbiche** [comment]
> ```
> inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-20.09";
> inputs.nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";
> 
> inputs.home-manager.url = "github:nix-community/home-manager/master";
> inputs.home-manager.inputs.nixpkgs.follows = "/nixpkgs-unstable";
> ```

The `nixpkgs.follows = "/nixpkgs-unstable"` has a leading "/" (error while copy-pasting?)

Home Manager doesn't have it's own `flake.lock`, so the `nixpkgs` revision it uses is not pinned.

What is the output of `nix flake list-inputs`?

> it's not clear to me where the input is going since there are no listed inputs for nixpkgs in the flake

There's a set of default flake inputs, it has something to do with a registry but I'm not really confident about my knowledge in that regard.
More information is available here: https://github.com/NixOS/nix/blob/a93916b1905cd7b968e92cd94a3e4a595bff2e0f/src/nix/registry.md

**Comment by stites** [comment]
Okay, I figured this out for my situation. The fix was just to call the home manager module with the expected arguments. 

```
({ config, utils, ... }: home.nixosModules.home-manager {
   pkgs = unstablePkgs;
   lib = unstablePkgs.lib;
   inherit config utils;
})
```
with `unstablePkgs` as:
```
unstablePkgs = import unstable {
  inherit system;
  config = { allowUnfree = true; };
};
```

it looks like I don't need to change my _module.args at all and I can enable octant (https://github.com/nix-community/home-manager/pull/1620/files) which is in master.

For context, `nix flake list-inputs` shows:

```
$ nix flake list-inputs              
warning: Git tree '/etc/nixos' is dirty
git+file:///etc/nixos
├───emacs: github:nix-community/emacs-overlay/82a9f26161e10691233cf757bddfd9d7c70ac940
├───home: github:rycee/home-manager/8127799f79ee96129b295d78294f40a54078131f
│   └───nixpkgs follows input 'unstable'
├───nixos: github:nixos/nixpkgs/e065200fc90175a8f6e50e76ef10a48786126e1c
├───nixos-hardware: github:NixOS/nixos-hardware/c242378e63b0ec334e964ac0c0fbbdd2b3e89ebf
├───nixpkgs: github:nixos/nixpkgs/c6b23ba64aea4b0e2163ff84990084a6bec455c5
├───nur: github:nix-community/NUR/db70e1906d253925740fc8f81ecc2b6d7cb6aab1
└───unstable: github:nixos/nixpkgs/56bb1b0f7a33e5d487dc2bf2e846794f4dcb4d01
```

**Comment by LU15W1R7H** [comment]
I tried to do what @stites proposed. But for some reason it doesn't work for me...
```nix
{
  description = "Personal NixOS Configuration";

  inputs = {
    nixpkgs.url = github:nixos/nixpkgs/nixos-unstable;
    nixpkgs-unstable.url = github:nixos/nixpkgs/nixpkgs-unstable;
    home-manager = {
      url = github:nix-community/home-manager;
      inputs.nixpkgs.follows = "nixpkgs-unstable";
    };
  };

  outputs = { self, nixpkgs, nixpkgs-unstable, home-manager }@attrs:
    let
      system = "x86_64-linux";
      hostname = "lwirth-tp";
      user = "luis";
      pkgs = import nixpkgs {
        inherit system;
        config.allowUnfree = true;
      };
      pkgs-unstable = import nixpkgs-unstable {
        inherit system;
        config.allowUnfree = true;
      };
    in
    {
      formatter.${system} = pkgs.nixpkgs-fmt;

      devShell.${system} = pkgs.mkShell {
        nativeBuildInputs = with pkgs; [
          rnix-lsp
        ];
        buildInputs = [ ];
      };

      nixosConfigurations.${hostname} = nixpkgs.lib.nixosSystem {
        inherit system;
        specialArgs = attrs;
        modules = [
          ./configuration.nix
          home-manager.nixosModules.home-manager
          {
            pkgs = pkgs-unstable;
            lib = pkgs-unstable.lib;
            home-manager.useGlobalPkgs = false;

            home-manager.useUserPackages = true;
            home-manager.users.${user} = {
              imports = [ ./home.nix ];
            };
          }
        ];
      };
    };
}
```
I get the following
```sh
> sudo nixos-rebuild switch --flake .
error: The option `pkgs' does not exist. Definition values:
       - In `/nix/store/l8p5y977qw6pi1iwd2mnfcn1rnil8cjr-source/flake.nix'
```

**Comment by stites** [comment]
I just ran into this recently as well and am just rolling back to 22.05 until next month. I suspect that the new flake integration will require a different way of doing this and taking into account the `stateVersion` will also be important. The intuitive thing didn't work, so I am punting for the time being.
